### PR TITLE
Moved ServiceConnect config from task to a new package

### PR DESCRIPTION
**Pull request by yinyic**
### Summary

This PR moves `ServiceConnect` config models and relevant parsing/validation logic into a separate `serviceconnect` package (originally under task package). This is to avoid a potential circular import during CNI plugin integration, where `task` imports `ecscni` and `ecscni` needs ServiceConnect config.

### Testing

New tests cover the changes:  no

### Description for the changelog

Moved ServiceConnect config from task to a new package

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/api/task/task.go [hunk 1]
```diff
@@ -281,7 +283,7 @@ type Task struct {
 	// setIdOnce is used to set the value of this task's id only the first time GetID is invoked
 	setIdOnce sync.Once
 
-	ServiceConnectConfig *ServiceConnectConfig `json:"ServiceConnectConfig,omitempty"`
+	ServiceConnectConfig *serviceconnect.Config `json:"Config,omitempty"`
```
Comment (aws-gibbskt) [inline agent/api/task/task.go:286]: I don't think we want to change the serialized key with this change.

File: agent/api/serviceconnect/service_connect_attachment_parser_test.go [hunk 2]
```diff
@@ -141,16 +143,16 @@ func getTestACSAttachments(attachmentProperties []*ecsacs.AttachmentProperty) *e
 	return &ecsacs.Attachment{
 		AttachmentArn:        strptr("attachmentArn"),
 		AttachmentProperties: attachmentProperties,
-		AttachmentType:       strptr(serviceConnectAttachmentType),
+		AttachmentType:       strptr(ServiceConnectAttachmentType),
```
Comment (aws-gibbskt) [inline agent/api/serviceconnect/service_connect_attachment_parser_test.go:146]: what is driving us to change encapsulation of this Type?

File: agent/api/task/task.go (hunk 1 above)
Comment (yinyic) [reply agent/api/task/task.go:286]: You are right this is rather unintended. Will fix.

File: agent/api/serviceconnect/service_connect_attachment_parser_test.go (hunk 2 above)
Comment (yinyic) [reply agent/api/serviceconnect/service_connect_attachment_parser_test.go:146]: Currently in `task` package when retrieving SC config we go through the list of task attachment and find the one that has `ServiceConnectAttachmentType`, before calling SC parser and validator (see [getServiceConnectConfig](https://github.com/aws/amazon-ecs-agent/blob/feature/service-connect/agent/api/task/task.go#L332)). 

I can create getter for it, or alternatively I can move the entire `getServiceConnectConfig` method from task to serviceconnect package.

Comment (aws-gibbskt) [reply agent/api/serviceconnect/service_connect_attachment_parser_test.go:146]: I think that aspect is very much a "task" thing more than a "serviceconnect" thing. So knowledge of the type name is probably "task" parsing a json blob or something could go in either place

Comment (yinyic) [reply agent/api/serviceconnect/service_connect_attachment_parser_test.go:146]: Makes sense. I'll move attachment type key back to `task`.

**Review by aws-gibbskt — DISMISSED** [review]

**Review by chienhanlin — DISMISSED** [review]

File: agent/api/serviceconnect/service_connect_attachment_parser.go [hunk 3]
```diff
@@ -23,34 +23,32 @@ import (
...
 
 const (
-	// serviceConnectAttachmentType specifies attachment type for service connect
-	serviceConnectAttachmentType = "ServiceConnect"
-	// serviceConnectConfigKey specifies the key maps to the service connect config in attachment properties
-	serviceConnectConfigKey = "ServiceConnectConfig"
-	// serviceConnectContainerNameKey specifies the key maps to the service connect container name in attachment properties
-	serviceConnectContainerNameKey = "ContainerName"
+	// ServiceConnectConfigKey specifies the key maps to the service connect config in attachment properties
+	ServiceConnectConfigKey = "Config"
```
Comment (rajal-amzn) [inline agent/api/serviceconnect/service_connect_attachment_parser.go:27]: The config key is coming in from Attachment and the value is still `ServiceConnectConfig`. Why is this changed?

**Review by rajal-amzn — CHANGES_REQUESTED** [review]

File: agent/api/serviceconnect/service_connect_attachment_parser.go (hunk 3 above)
Comment (yinyic) [reply agent/api/serviceconnect/service_connect_attachment_parser.go:27]: Sorry that's a miss. Will fix.

**Review by rajal-amzn — DISMISSED** [review]

File: agent/api/task/task_test.go [hunk 4]
```diff
@@ -4272,11 +4287,11 @@ func TestTaskServiceConnectAttachment(t *testing.T) {
 						AttachmentArn: strptr("attachmentArn"),
 						AttachmentProperties: []*ecsacs.AttachmentProperty{
 							{
-								Name:  strptr(serviceConnectConfigKey),
+								Name:  strptr(serviceconnect.ServiceConnectConfigKey),
 								Value: strptr(tc.testSCConfigValue),
 							},
 							{
-								Name:  strptr(serviceConnectContainerNameKey),
+								Name:  strptr(serviceconnect.ServiceConnectContainerNameKey),
```
Comment (aws-gibbskt) [inline agent/api/task/task_test.go:4294]: This test usage feels like a week reason to adjust visibility of these values with the s vs S beginning

Comment (yinyic) [reply agent/api/task/task_test.go]: Added getters. I thought about hardcoding them in this test file but I'm not sure if I like the idea that changing the internal service connect config keys will break this unit test when it shouldn't really matter.

Comment (aws-gibbskt) [reply agent/api/task/task_test.go]: That works, or you can have a function accessible in the test realm. Specifically `service_connect_attachment_parser_test.go` can have a helper `CreateServiceConnectAttachment` or something with arguments that can be passed in and returning a correctly key'd map to merge here. That function only exists in test and doesn't require getters or exporting and can provide a reminder when changing things in the attachment parser that you have to support that use-case

Comment (yinyic) [reply agent/api/task/task_test.go]: I couldn't seem to get it to work. It appears that test files are compiled separately for each package. According to https://pkg.go.dev/cmd/go/internal/test -`"Each listed package causes the execution of a separate test binary."` For that reason, test files could not access even exported functions from a test file in another package.

Comment (aws-gibbskt) [reply agent/api/task/task_test.go]: TIL - thanks.

**Review by aws-gibbskt — APPROVED** [review]

**Review by rajal-amzn — APPROVED** [review]

**Review by chienhanlin — APPROVED** [review]

**Review by ubhattacharjya — APPROVED** [review]
